tests/src/Mod/Part/App/TopoShapeMakeShapeWithElementMap.cpp

```diff
@@ -48,14 +55,107 @@ class TopoShapeMakeShapeWithElementMapTests: public ::testing::Test
     std::vector<Part::TopoShape> sources {cube1, cube2};
-    TopoDS_Vertex nullShape;
+    TopoDS_Vertex nullVertex;
+    TopoDS_Edge nullEdge;
+    TopoDS_Wire nullWire;
+    TopoDS_Face nullFace;
+    TopoDS_Shell nullShell;
+    TopoDS_Solid nullSolid;
+    TopoDS_CompSolid nullCompSolid;
+    TopoDS_Compound nullCompound;
 
     // Act and assert
-    EXPECT_THROW(Shape()->makeShapeWithElementMap(nullShape, *Mapper(), sources),
+    EXPECT_THROW(Shape()->makeShapeWithElementMap(nullVertex, *Mapper(), sources),
                  Part::NullShapeException);
+    EXPECT_THROW(Shape()->makeShapeWithElementMap(nullEdge, *Mapper(), sources),
+                 Part::NullShapeException);
+    EXPECT_THROW(Shape()->makeShapeWithElementMap(nullWire, *Mapper(), sources),
+                 Part::NullShapeException);
+    EXPECT_THROW(Shape()->makeShapeWithElementMap(nullFace, *Mapper(), sources),
+                 Part::NullShapeException);
+    EXPECT_THROW(Shape()->makeShapeWithElementMap(nullShell, *Mapper(), sources),
+                 Part::NullShapeException);
+    EXPECT_THROW(Shape()->makeShapeWithElementMap(nullSolid, *Mapper(), sources),
+                 Part::NullShapeException);
+    EXPECT_THROW(Shape()->makeShapeWithElementMap(nullCompSolid, *Mapper(), sources),
+                 Part::NullShapeException);
+    EXPECT_THROW(Shape()->makeShapeWithElementMap(nullCompound, *Mapper(), sources),
+                 Part::NullShapeException);
+}
+
+TEST_F(TopoShapeMakeShapeWithElementMapTests, emptySourceShapes)
+{
+    // Arrange
+    auto [cube1, cube2] = TopoShapeExpansionHelpers::CreateTwoCubes();
+    std::vector<Part::TopoShape> emptySources;
+    std::vector<Part::TopoShape> nonEmptySources {cube1, cube2};
+
+    // Act and assert
+    for (auto& source : nonEmptySources) {
+        Part::TopoShape& modifiedShape = source;
+        Part::TopoShape& originalShape = source;
+
+        EXPECT_EQ(
+            &originalShape,
+            &modifiedShape.makeShapeWithElementMap(source.getShape(), *Mapper(), emptySources));
+    }
+}
+
+TEST_F(TopoShapeMakeShapeWithElementMapTests, nonMappableSources)
+{
+    // Arrange
+    auto [cube1, cube2] = TopoShapeExpansionHelpers::CreateTwoCubes();
+    std::vector<Part::TopoShape> sources {cube1, cube2};
+
+    // Act and assert
+    for (auto& source : sources) {
+        size_t canMap = 0;
+        for (const auto& mappableSource : sources) {
+            if (source.canMapElement(mappableSource)) {
+                ++canMap;
+            }
+        }
+
+        if (canMap == 0U) {
+            EXPECT_EQ(&source,
+                      &source.makeShapeWithElementMap(source.getShape(), *Mapper(), sources));
+        }
+    }
+}
+
+TEST_F(TopoShapeMakeShapeWithElementMapTests, findShapeInElementMap)
+{
+    // Arrange
+    auto [cube1, cube2] = TopoShapeExpansionHelpers::CreateTwoCubes();
+    std::vector<Part::TopoShape> sources {cube1, cube2};
+    sources[0].Tag = 1;
+    sources[1].Tag = 2;
```

Comment:
Are these required?  If so, maybe a line comment saying something like :  // Untagged shapes don't map

Reply:
Yes, otherwise `makeShapeWithElementMap()` will return at the same point of the test `nonMappableSources`. However you're right, better making this explicit with a comment!